### ISSUE TITLE
Monitor hangs in `SerialMonitor.serial_write()` if the application never reads the UART (IDFGH-10821)

### ISSUE THREAD

**Issue by nomis** [espressif/esp-idf-monitor#3]
My application doesn't do anything with data written to the UART. It won't initialise it or read from it.

After typing a few characters the output buffer becomes full and the monitor then hangs in `SerialMonitor.serial_write()`. The only feature that still works is the help output. Everything else (including exit) requires the command queue but the main loop is stuck waiting for the serial write to complete.

It's easy to accidentally press characters without Ctrl so the monitor regularly hangs and I have to kill it from another window. I then need to run `reset` to fix my terminal and lose the scrollback buffer.

I'm using Linux 5.4 with the serial port of the ESP32-C6-DevKitC-1 V1.2 "USB" connection:
```
New USB device found, idVendor=303a, idProduct=1001, bcdDevice= 1.02
New USB device strings: Mfr=1, Product=2, SerialNumber=3
Product: USB JTAG/serial debug unit
Manufacturer: Espressif
```

The serial writes need to happen in a background thread that can be stopped by closing the serial port. That would allow flow control to remain enabled (e.g. if someone is pasting lots of text).

**Comment by dobairoland** [comment]
I think this might be the same issue as https://github.com/espressif/esp-idf/issues/11896. We already have a possible solution under testing for it.

**Comment by nomis** [comment]
Any change to the device would not fix this bug in the monitor.

There's a `select()` call waiting for the serial port to become writeable and it needs to handle that properly.

**Comment by nomis** [comment]
How do I reopen the issue? That commit doesn't properly fix the problem. It still hangs for 300ms **per character** and you're incrementing a variable that hasn't been set yet:

```
    self.timeout_cnt += 1
AttributeError: 'SerialMonitor' object has no attribute 'timeout_cnt'
```

**Comment by nomis** [comment]
If I type 4 characters (all of which will block and I get the warning message once) and then `^T^X`, the monitor takes 30 seconds to exit. During that time anything I type (including `^T` etc.) is echoed immediately.

e.g. (assuming I initialise `self.timeout_cnt = 0` properly):

```
Writing to serial is timing out. Please make sure that your application supports an interactive console and that you have picked the correct console for serial communication.
asdf
```

```
Writing to serial is timing out. Please make sure that your application supports an interactive console and that you have picked the correct console for serial communication.
^T^H^T^X^T^X^X^T^X^T^X^T^X
```

I already tried the `write_timeout` approach myself and found it didn't work.

**Comment by peterdragun** [comment]
>  you're incrementing a variable that hasn't been set yet:
I am sorry for that, seems like it always passed first time for me (probably buffer was 1) so the variable was set during testing, I will fix it.

> It still hangs for 300ms per character
Yes it does to make sure that write was really not successful. We are trying to provide solution for everyone no matter if they are aware of new cli option or not, so the app won't hang in write. If you have an app that does not support writing to serial console you will get a warning that you are doing something that is not supported. If you continue writing characters to that console after the warning there is nothing to do from the monitor side.

> the monitor takes 30 seconds to exit. During that time anything I type (including ^T etc.) is echoed immediately.
Monitor is waiting for OS to close the connection. We are aware of the issue.

**Comment by nomis** [comment]
> We are trying to provide solution for everyone no matter if they are aware of new cli option or not, so the app won't hang in write.

The warning message should be output for the **first** discarded character. Otherwise you may be losing data without warning for some people based solely on a short write timeout.

> If you have an app that does not support writing to serial console you will get a warning that you are doing something that is not supported. If you continue writing characters to that console after the warning there is nothing to do from the monitor side.

The correct solution is still to fix the monitor so that it doesn't hang regardless of what the device does. Even if it supports reading from the serial console, if the device freezes when someone pastes 1000 characters the monitor will hang for 5 minutes being unable to exit or reset the device (at least until a device watchdog resets the device).

**Comment by nomis** [comment]
The closing wait time should be set to 0 seconds if there has been a write failure.

It also doesn't work for me:
```
Exception in thread Thread-2:
Traceback (most recent call last):
  File "/usr/lib/python3.8/threading.py", line 932, in _bootstrap_inner
    self.run()
  File "/usr/lib/python3.8/threading.py", line 870, in run
    self._target(*self._args, **self._kwargs)
  File "/home/simon/.espressif/python_env/idf5.1_py3.8_env/lib/python3.8/site-packages/esp_idf_monitor/base/stoppable_thread.py", line 47, in _run_outer
    self.run()
  File "/home/simon/.espressif/python_env/idf5.1_py3.8_env/lib/python3.8/site-packages/esp_idf_monitor/base/serial_reader.py", line 100, in run
    self.close_serial()
  File "/home/simon/.espressif/python_env/idf5.1_py3.8_env/lib/python3.8/site-packages/esp_idf_monitor/base/serial_reader.py", line 105, in close_serial
    self.set_closing_wait(delay_sec=1)
  File "/home/simon/.espressif/python_env/idf5.1_py3.8_env/lib/python3.8/site-packages/esp_idf_monitor/base/serial_reader.py", line 128, in set_closing_wait
    fcntl.ioctl(self.serial.fd, termios.TIOCSSERIAL, buf)
PermissionError: [Errno 1] Operation not permitted
```

You can't set `closing_wait` unless the process has `CAP_SYS_ADMIN`,  which is only likely to be the case if you're running the monitor as root:
```
        if (!capable(CAP_SYS_ADMIN)) {
                if ((close_delay != acm->port.close_delay) ||
                    (closing_wait != acm->port.closing_wait))
                        retval = -EPERM;
        } else {
                acm->port.close_delay  = close_delay;
                acm->port.closing_wait = closing_wait;
        }
```

**Comment by nomis** [comment]
I've tried `termios.tcflush()` (it's also possible to use `termios.tcflow()` to suspend output before closing) but the `cdc-adm` driver doesn't implement the `flush_buffer` tty operation so it has no effect.

I can't see any way around this without fixing `cdc-adm`. Even if I just discard that fd the kernel will close it when the process exits.

I have submitted a patch to implement `flush_buffer` in `cdc-adm`.
